packages/adblocker/src/engine/bucket/preprocessor.ts

```diff
@@ -0,0 +1,157 @@
+import { StaticDataView } from '../../data-view';
+import IFilter from '../../filters/interface';
+import Preprocessor, { Env } from '../../preprocessor';
+
+export default class PreprocessorBucket {
+  public static deserialize(view: StaticDataView): PreprocessorBucket {
+    const excluded = new Set<number>();
+    for (let i = 0, l = view.getUint32(); i < l; i++) {
+      excluded.add(view.getUint32());
+    }
+
+    const preprocessors: Preprocessor[] = [];
+    for (let i = 0, l = view.getUint32(); i < l; i++) {
+      preprocessors.push(Preprocessor.deserialize(view));
+    }
+
+    return new this({
+      excluded,
+      preprocessors,
+    });
+  }
+
+  private readonly preprocessors: Preprocessor[];
+  private readonly excluded: Set<number>;
+
+  constructor({
+    excluded = new Set(),
+    preprocessors = [],
+  }: {
+    excluded?: Set<number>;
+    preprocessors?: Preprocessor[];
+  }) {
+    this.excluded = excluded;
+    this.preprocessors = preprocessors;
+  }
+
+  private build(env: Env) {
+    // Update excluded filter ids based on bindings
+    this.excluded.clear();
+
+    const included = new Set<number>();
+
+    for (let i = 0; i < this.preprocessors.length; ) {
+      if (this.preprocessors[i].filterIDs.size === 0) {
+        this.preprocessors.splice(i, 1);
+
+        continue;
+      }
+
+      if (this.preprocessors[i].evaluate(env)) {
+        for (const filterID of this.preprocessors[i].filterIDs) {
+          included.add(filterID);
+        }
+      } else {
+        for (const filterID of this.preprocessors[i].filterIDs) {
+          if (!included.has(filterID)) {
```

Comment:
It'll drop the filter in the reverse case:

```
!#if true
filter
!#endif
!#if false
filter
!#endif
```

We should include the filter if any of filter condition is met.

Reply:
That's exactly what I'm not sure of. I would say it should drop a filter if at least one condition is not met :)

Reply:
I agree with @chrmod and this would make the implementation much simpler as well.

Reply:
I think it'll create an unexpected behavior because it's not intended. For example, we don't know if the filter maintainers will write like:

```
||domain

!#if platform_a
@@||domain
!#endif

!#if platform_b
@@||domain
!#endif
```

We cannot ensure the input always will be like:

```
||domain

!#if platform_a || platform_b
@@||domain
!#endif
```

Reply:
Behavior changed :)